core/templates/dev/head/services/CsrfTokenService.ts

```diff
@@ -18,33 +18,39 @@
 
 oppia.factory('CsrfTokenService', [function() {
   var token = null;
+  var tokenPromise = null;
+
   return {
-    setToken: function(csrfToken) {
-      token = csrfToken;
+    initializeToken: function() {
+      if (tokenPromise !== null) {
+        throw new Error('Token request already made');
+      }
+      tokenPromise = new Promise(function(resolve, reject) {
+        // We use jQuery here instead of Angular's $http, since the latter
+        // creates a circular dependency.
+        $.ajax({
+          url: '/csrfhandler',
+          type: 'GET',
+          dataType: 'text',
+          dataFilter: function(data) {
+            // Remove the XSSI prefix.
+            var transformedData = data.substring(5);
+            return JSON.parse(transformedData);
+          },
+        }).done(function(response) {
+          token = response.token;
+          resolve(response.token);
+        });
+      });
     },
+
     getToken: function() {
-      // We use jQuery here instead of Angular's $http, since the latter
-      // creates a circular dependency.
-      return new Promise(function(resolve, reject) {
-        if (token) {
+      if (token) {
```

Comment:
Should we throw an error if no token is set (i.e. initialization has not happened yet)? If you think it's possible for this to be called before initialization, then change the name to getTokenAsync(), and add a test that calls getToken() before initializeToken() and make sure that works correctly. Otherwise, throw an error if token is null and just return the token otherwise; no need to return a promise.

Reply:
I think we'll still need to return a promise because it is possible that the initialization is not complete when a POST or PUT request is sent so the request will have to wait for the initialization to be complete.

Reply:
Done PTAL

Reply:
OK, can you add a test that calls getToken() before initializeToken() so that we can be sure this works correctly in the case where token initialization has some latency?

Reply:
I've done that already [here](https://github.com/oppia/oppia/pull/6951#discussion_r298698635)